omnigibson/robots/behaviorbot.py

```diff
@@ -0,0 +1,677 @@
+        # We have to use delicate logic to account for the edge cases where a single joint may contain > 1 DOF
+        # (e.g.: spherical joint)
+        cur_indices_idx = 0
+        while cur_indices_idx != n_indices:
+            # Grab the current DOF index we're controlling and find the corresponding joint
+            joint = self._dof_to_joints[indices[cur_indices_idx]]
+            cur_ctrl_idx = indices[cur_indices_idx]
+            joint_dof = joint.n_dof
+            if joint_dof > 1:
+                # Run additional sanity checks since the joint has more than one DOF to make sure our controls,
+                # control types, and indices all match as expected
+
+                # Make sure the indices are mapped correctly
+                assert indices[cur_indices_idx + joint_dof] == cur_ctrl_idx + joint_dof, \
+                    "Got mismatched control indices for a single joint!"
+                # Check to make sure all joints, control_types, and normalized as all the same over n-DOF for the joint
+                for group_name, group in zip(
+                        ("joints", "control_types", "normalized"),
+                        (self._dof_to_joints, control_type, normalized),
+                ):
+                    assert len({group[indices[cur_indices_idx + i]] for i in range(joint_dof)}) == 1, \
+                        f"Not all {group_name} were the same when trying to deploy control for a single joint!"
+                # Assuming this all passes, we grab the control subvector, type, and normalized value accordingly
+                ctrl = control[cur_ctrl_idx: cur_ctrl_idx + joint_dof]
+            else: 
+                # Grab specific control. No need to do checks since this is a single value
+                ctrl = control[cur_ctrl_idx]
+
+            # Deploy control based on type
+            ctrl_type, norm = control_type[cur_ctrl_idx], normalized[cur_ctrl_idx]       # In multi-DOF joint case all values were already checked to be the same
+            if ctrl_type == ControlType.EFFORT:
+                joint.set_effort(ctrl, normalized=norm)
+            elif ctrl_type == ControlType.VELOCITY:
+                joint.set_vel(ctrl, normalized=norm, drive=True)
+            elif ctrl_type == ControlType.POSITION:            
+                if "rx" in joint.joint_name or "ry" in joint.joint_name or "rz" in joint.joint_name: 
+                    joint.set_pos(ctrl, normalized=norm, drive=False)
+                else:
+                    joint.set_pos(ctrl, normalized=norm, drive=True)
+            elif ctrl_type == ControlType.NONE:
+                # Do nothing
+                pass
+            else:
+                raise ValueError("Invalid control type specified: {}".format(ctrl_type))
+
+            # Finally, increment the current index based on how many DOFs were just controlled
+            cur_indices_idx += joint_dof
+
+    def teleop_data_to_action(self, teleop_data: dict) -> np.ndarray:
```

Comment:
add docstring: what should teleop_data consist of?

Reply:
Added a pointer to the docstring in TeleopSystem class